### DISCUSSION TITLE
HID Support

### DISCUSSION THREAD

**Discussion by jwillikers** [adafruit/Adafruit_nRF52_Bootloader#239] [Ideas]
**Is your feature request related to a problem? Please describe.**

I'd like to use the [hf2-rs](https://github.com/jacobrosenthal/hf2-rs) utility to develop in Rust for the Circuit Playground Bluefruit, like I currently do for the Circuit Playground Express.
This is not possible right now because the nRF52 bootloader does not support _HID_, which is necessary.
The samd bootloader, on the other hand, has this support available.

**Describe the solution you'd like**

I'd like Adafruit's nRF52 bootloader to add HID support like that available in the [uf2-samdx1](https://github.com/adafruit/uf2-samdx1) bootloader.

**Describe alternatives you've considered**

Manually creating a _UF2_ firmware file and using drag-and-drop or manually creating a DFU archive and uploading that with `adafruit-nrfutil`.

**Additional context**

The original issue is jacobrosenthal/hf2-rs#41.
HID support appears to mostly be implemented in [uf2hid.h](https://github.com/adafruit/uf2-samdx1/blob/master/inc/uf2hid.h), [uf2.h](https://github.com/adafruit/uf2-samdx1/blob/master/inc/uf2.h), and [hid.c](https://github.com/adafruit/uf2-samdx1/blob/master/src/hid.c) in uf2-samdx1.

**Comment by ladyada** [comment]
hiya - if you submit a PR we could review and merge it!

**Comment by TheButlah** [comment]
Hi! I actually have the exact same use case. HF2 support would make flashing the board a lot easier. This is because in my case, I cannot seem to get the ELF files that rustc generates converted into bin files, and then converting from bin to DF2 files is another step.

Supporting HF2 would also allow for flashing via WebHID, which is another use case to consider.

I would offer to submit a PR but unfortunately this whole rabbit hole is my first ever experience with flashing firmware. I don't think I have the necessary experience to devise a PR.

More information is available at https://github.com/microsoft/uf2/blob/master/hf2.md

**Reply by jwillikers** [reply]
@TheButlah If it helps, you can check out the [README](https://github.com/jwillikers/smart-leds-fx/tree/main/boards/circuit_playground_bluefruit#program) in my project for how to do the conversion manually for the time being.

**Reply by TheButlah** [reply]
~~When  I do this, the .bin file is empty. I'm not really sure why - my main.rs is not empty~~

Fixed the empty .bin - my memory.x and linker arguments needed to be specified.
